pandas/_libs/tslib.pyx

```diff
@@ -217,38 +219,48 @@ def array_with_unit_to_datetime(
 
     assert is_ignore or is_coerce or is_raise
 
-    if unit == 'ns':
-        if issubclass(values.dtype.type, np.integer):
-            result = values.astype('M8[ns]')
+    if unit == "ns":
+        if issubclass(values.dtype.type, (np.integer, np.float_)):
+            result = values.astype("M8[ns]")
```

Comment:
does astype of floats directly to M8 work?

Reply:
i guess it does (as you use it below), but do we have a test  specifically for float with unit='ns'? 

also can try .astype(..., copy=False)

Reply:
> does astype of floats directly to M8 work?

would it be better to do, here and below:
``` python
ivalues = values.view("i8")
result = ivalues.astype("M8[ns]")
```

Reply:
> i guess it does (as you use it below), but do we have a test specifically for float with unit='ns'?

I'll look some more but I *think* we don't. Will add unless I find one